runtime/libc/wasi_sdk_backing.c

```diff
@@ -0,0 +1,500 @@
+        case ENOPROTOOPT:       return WASI_ENOPROTOOPT;
+        case ENOSPC:            return WASI_ENOSPC;
+        case ENOSYS:            return WASI_ENOSYS;
+        case ENOTCONN:          return WASI_ENOTCONN;
+        case ENOTDIR:           return WASI_ENOTDIR;
+        case ENOTEMPTY:         return WASI_ENOTEMPTY;
+        case ENOTRECOVERABLE:   return WASI_ENOTRECOVERABLE;
+        case ENOTSOCK:          return WASI_ENOTSOCK;
+        case ENOTSUP:           return WASI_ENOTSUP;
+        case ENOTTY:            return WASI_ENOTTY;
+        case ENXIO:             return WASI_ENXIO;
+        case EOVERFLOW:         return WASI_EOVERFLOW;
+        case EOWNERDEAD:        return WASI_EOWNERDEAD;
+        case EPERM:             return WASI_EPERM;
+        case EPIPE:             return WASI_EPIPE;
+        case EPROTO:            return WASI_EPROTO;
+        case EPROTONOSUPPORT:   return WASI_EPROTONOSUPPORT;
+        case EPROTOTYPE:        return WASI_EPROTOTYPE;
+        case ERANGE:            return WASI_ERANGE;
+        case EROFS:             return WASI_EROFS;
+        case ESPIPE:            return WASI_ESPIPE;
+        case ESRCH:             return WASI_ESRCH;
+        case ESTALE:            return WASI_ESTALE;
+        case ETIMEDOUT:         return WASI_ETIMEDOUT;
+        case ETXTBSY:           return WASI_ETXTBSY;
+        case EXDEV:             return WASI_EXDEV;
+    }
+}
+
+// file operations
+i32 wasi_unstable_path_open(
+        i32 dirfd,
+        u32 lookupflags,
+        u32 path_off,
+        u32 path_len,
+        u32 oflags,
+        u64 fs_rights_base,
+        u64 fs_rights_inheriting,
+        u32 fdflags,
+        u32 fd_off) {
+    // get path
+    char* path = get_memory_string(path_off);
+
+    // translate o_flags and fs_flags into flags and mode
+    int flags = (
+            ((oflags & WASI_O_CREAT    ) ? O_CREAT     : 0) |
+            ((oflags & WASI_O_DIRECTORY) ? O_DIRECTORY : 0) |
+            ((oflags & WASI_O_EXCL     ) ? O_EXCL      : 0) |
+            ((oflags & WASI_O_TRUNC    ) ? O_TRUNC     : 0) |
+            ((fdflags & WASI_FDFLAG_APPEND  ) ? O_APPEND   : 0) |
+            ((fdflags & WASI_FDFLAG_DSYNC   ) ? O_DSYNC    : 0) |
+            ((fdflags & WASI_FDFLAG_NONBLOCK) ? O_NONBLOCK : 0) |
+            ((fdflags & WASI_FDFLAG_RSYNC   ) ? O_RSYNC    : 0) |
+            ((fdflags & WASI_FDFLAG_SYNC    ) ? O_SYNC     : 0));
+    if ((fs_rights_base & WASI_RIGHT_FD_WRITE) &&
+        (fs_rights_base & WASI_RIGHT_FD_READ)) {
+        flags |= O_RDWR;
+    } else if (fs_rights_base & WASI_RIGHT_FD_WRITE) {
+        flags |= O_WRONLY;
+    } else if (fs_rights_base & WASI_RIGHT_FD_READ) {
+        flags |= O_RDONLY;
+    }
+
+    int mode = 0644;
+    int fd = openat(dirfd, path, flags, mode);
+    if (fd < 0) {
+        return wasi_fromerrno(errno);
+    }
+
+    set_i32(fd_off, fd);
+    return WASI_ESUCCESS;
+}
+
+i32 wasi_unstable_fd_close(i32 fd) {
```

Comment:
So undefined symbols go to "env" but wasi explicitly imports from a different namespace. I didn't even realize my existing jank would handle that case 😃